File: q2598/q2598.java
class Solution {
    public int findSmallestInteger(int[] nums, int value) {
        Map<Integer, Integer> remainderCount = new HashMap<>();
        for (int num : nums) {
            int remainder = ((num % value) + value) % value;
            remainderCount.put(remainder, remainderCount.getOrDefault(remainder, 0) + 1);
        }

        int start = 0; 
        while (remainderCount.getOrDefault(start % value, 0) > 0) {
            remainderCount.put(start % value, remainderCount.get(start % value) - 1);
            start++;
        }
        return start; 
    }
}

File: q2598/q2598.py
class Solution:
    def findSmallestInteger(self, nums: List[int], value: int) -> int:
        remainderCount = defaultdict(int)
        for num in nums: 
            remainder = num % value
            remainderCount[remainder] += 1
        
        # Try to build MEX starting from 0 upwards
        start = 0
        while remainderCount[start % value] > 0:
            # If we have at least one number of this remainder class
            remainderCount[start % value] -= 1  # Use it to represent `start`
            start += 1
        
        # Can't represent this number, so MEX is `start`
        return start
